source/extensions/common/workload_discovery/api.cc

```diff
@@ -138,7 +138,9 @@ class WorkloadMetadataProviderImpl : public WorkloadMetadataProvider, public Sin
             dynamic_cast<const istio::workload::Address&>(resource.get().resource());
         switch (address.type_case()) {
         case istio::workload::Address::kWorkload:
-          index->emplace(address.workload().address(), convert(address.workload()));
+          for (const auto& addr : address.workload().addresses()) {
+            index->emplace(addr, convert(address.workload()));
```

Comment:
Doesn't this prevent lookup-by-address for the on-demand mode?
The xDS resource key must be the lookup key.

Reply:
xDS supports primary and alias keys: https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#resource-names

we can still lookup on demand by any of the alias keys (the IP addresses) but the primary key is a single primary name we can depend on (since which IP address to use in e.g. dual stack is confusing)

Reply:
Are you sure about that? Last time I looked at the code, if a client requests resource X it has to receive resource X with possible alias Y. If a client request Y, then it gets Y with an alias X. There's a lot of diffing logic that may not handle a different key from what is requested.

Reply:
it's my understanding of @howardjohn's proposal here https://github.com/istio/istio/pull/45170#discussion_r1207338408

we still need to minorly update ztunnel / istio to handle this, but this is the plan

Reply:
It could be a limitation of the implementation right now. In any case, we're not doing on-demand right now so this can be deferred.